File: Day_4.R
extractInfo <- function(filepath) {
  con <- file(filepath, "r")
  all_info <- c()
  info <- c()
  while (TRUE) {
    line <- readLines(con, n = 1)
    if ( length(line) != 0 ) {
      if ( line != "" ) {
        info <- paste(info, line, sep = " ")
      } else {
        all_info <- rbind(all_info, info)
        info <- c()
      }
    } else {
      all_info <- rbind(all_info, info)
      break
    }
  }
  close(con)
  all_info
}

validPassport <- function(filepath, fields) {
  num <- 0
  pp_info <- extractInfo(filepath)
  for(i in 1:nrow(pp_info)) {
    valid <- c()
    for ( field in fields ) {
      exist <- grepl(field, pp_info[i, ])
      valid <- c(valid, exist)
    }
    if ( all(valid) == TRUE ) num <- num + 1
  }
  num
}

data_path <- "./Data/day_4.txt"

fields_p1 <- c("byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid")

fields_p2 <- c(
  "byr:19[2-9]\\d|200[0-2]",
  "iyr:20(1\\d|20)",
  "eyr:20(2\\d|30)",
  "hcl:#[0-9a-f]{6}",
  "ecl:(amb|blu|brn|gry|grn|hzl|oth)",
  "pid:\\d{9}( |$)",
  "hgt:(((1[5-8]\\d|19[0-3])cm)|((59|6\\d|7[0-6])in))"
)

cat("Part_1: # of valid passports:", validPassport(data_path, fields_p1))
cat("Part_2: # of valid passports:", validPassport(data_path, fields_p2))

File: day4.py
import re

require_fields = ["byr:", "iyr:", "eyr:", "hgt:", "hcl:", "ecl:", "pid:"]

with open("day4.in") as f:
    lines = f.read().split("\n")
lens = len(lines)

current = " "
passports = []
for line in lines:
    if len(line) == 0:
        passports.append(current)
        current = " "
    else:
        current = current + " " + line
passports.append(current)

print("tottal number of passport: ", len(passports))


def passport_check(passport):
    for require in require_fields:
        if require not in passport:
            return 0
    return 1


valid_passports = 0
for passport in passports:
    valid_passports += passport_check(passport)

print("valid number of passport: ", valid_passports)


def strict_check(passport):
    content = {
        x.split(":")[0]: x.split(":")[1] for x in passport.split(" ") if x.strip()
    }

    if (
        not re.match(r"^\d{4}$", content["byr"])
        or int(content["byr"]) < 1920
        or int(content["byr"]) > 2002
    ):
        return 0
    if (
        not re.match(r"^\d{4}$", content["iyr"])
        or int(content["iyr"]) < 2010
        or int(content["iyr"]) > 2020
    ):
        return 0
    if (
        not re.match(r"^\d{4}$", content["eyr"])
        or int(content["eyr"]) < 2020
        or int(content["eyr"]) > 2030
    ):
        return 0

    match = re.match(r"^(\d+)(cm|in)$", content["hgt"])
    if not match:
        return 0
    elif match.group(2) == "cm" and (
        int(match.group(1)) < 150 or int(match.group(1)) > 193
    ):
        return 0
    elif match.group(2) == "in" and (
        int(match.group(1)) < 59 or int(match.group(1)) > 76
    ):
        return 0

    if not re.match(r"^#[0-9a-f]{6}$", content["hcl"]):
        return 0
    if content["ecl"] not in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]:
        return 0
    if not re.match(r"^\d{9}$", content["pid"]):
        return 0
    return 1


valid_passports = 0
count = 0
for passport in passports:
    if passport_check(passport) and strict_check(passport):
        valid_passports += 1

    count += 1
print("valid passports strict role: ", valid_passports)